rope/contrib/autoimport/parse.py

```diff
@@ -0,0 +1,135 @@
+        return NameType.Function
+    if isinstance(node, ast.Assign):
+        return NameType.Variable
+    return NameType.Text  # default value
+
+
+def get_names_from_file(
+    module: pathlib.Path,
+    underlined: bool = False,
+) -> Generator[PartialName, None, None]:
+    """Get all the names from a given file using ast."""
+    with open(module, mode="rb") as file:
+        try:
+            root_node = ast.parse(file.read())
+        except SyntaxError as error:
+            print(error)
+            return
+    for node in ast.iter_child_nodes(root_node):
+        if isinstance(node, ast.Assign):
+            for target in node.targets:
+                try:
+                    assert isinstance(target, ast.Name)
+                    if underlined or not target.id.startswith("_"):
+                        yield PartialName(
+                            target.id,
+                            get_type_ast(node),
+                        )
+                except (AttributeError, AssertionError):
+                    # TODO handle tuple assignment
+                    pass
+        elif isinstance(node, (ast.FunctionDef, ast.ClassDef)):
+            if underlined or not node.name.startswith("_"):
+                yield PartialName(
+                    node.name,
+                    get_type_ast(node),
+                )
+
+
+def get_type_object(imported_object) -> NameType:
+    """Determine the type of an object."""
+    if inspect.isclass(imported_object):
+        return NameType.Class
+    if inspect.isfunction(imported_object) or inspect.isbuiltin(imported_object):
+        return NameType.Function
+    return NameType.Constant
+
+
+def get_names(module: ModuleInfo, package: Package) -> List[Name]:
+    """Get all names from a module and package."""
+    if isinstance(module, ModuleCompiled):
+        return list(
+            get_names_from_compiled(package.name, package.source, module.underlined)
+        )
+    if isinstance(module, ModuleFile):
+        return [
+            combine(package, module, partial_name)
+            for partial_name in get_names_from_file(module.filepath, module.underlined)
+        ]
+    return []
+
+
+def get_names_from_compiled(
+    package: str,
+    source: Source,
+    underlined: bool = False,
+) -> Generator[Name, None, None]:
+    """
+    Get the names from a compiled module.
+
+    Instead of using ast, it imports the module.
+    Parameters
+    ----------
+    package : str
+        package to import. Must be in sys.path
+    underlined : bool
+        include underlined names
+    """
+    # builtins is banned because you never have to import it
+    # python_crun is banned because it crashes python
+    banned = ["builtins", "python_crun"]
+    if package in banned or (package.startswith("_") and not underlined):
+        return  # Builtins is redundant since you don't have to import it.
+    try:
+        module = import_module(str(package))
```

Comment:
I think compiled-only packages probably should not be indexed by default. `import`-ing a compiled python module involves executing executable code from the imported module's, which means it may include running potentially harmful or malicious code.

Unless we can statically analyze the compiled modules somehow, this should not be default behavior.

At the very least, I'd suggest removing this code from `get_names()` for now or restricting it to standard library modules and re-think on how to handle compiled modules as a separate ticket/PR.

Reply:
For now I limited importing compiled modules to Project, Standard, and Builtin modules which isn't 100% accurate. I can remove Project or Standard, but builtin has to remain for modules like os and sys.
I agree that not importing modules is a good idea, but we do need it  unless we want to bundle a python decompiler.

Reply:
Done. It doesn't really make sense to keep Project since it'd require the user to have a compiled .so module in their local project *without* the python code. In the future, we could add a whitelist or something for these edge cases.